File: project_01/Dior.py
n = int(input())
b = input().split()
a = [0]*n
for i in range(len(b)):
    a[i] = int(b[i])

num = {}
bb = input().split()
q = [0]*n
for i in range(n):
    q[i] = int(bb[i])

a.sort()
for i in range(n):
    num[a[i]] = i

ans = 0
index = -1


def solve(p, r):
    global index
    if r - p >= 1:
        index += 1
    if r - p <= 1:
        return
    global ans
    ans += (r-p-1)
    pivot = q[index]
    id = num[pivot]
    solve(p, id)
    solve(id+1, r)


solve(0, n)
print(ans)

File: project_01/Dior.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Scanner;

public class Dior {

    static int index = -1;
    static int[] a;
    static Long ans = 0L;
    static int[] pivot_list;
    static HashMap<Integer, Integer> hashMap;
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        a = new int[n];
        for(int i = 0; i < n; i++){
            a[i] = scanner.nextInt();
        }
        pivot_list = new int[n];
        for(int i = 0; i < n; i++){
            pivot_list[i] = scanner.nextInt();
        }

        Arrays.sort(a);
        hashMap = new HashMap<>();
        for(int i = 0; i < n; i++) {
            hashMap.put(a[i], i);
        }

        solve(0, n);
        System.out.print(ans);

    }

    static void solve(int p, int r){
        if(r-p >= 1){
           index++;
        }
        if(r-p <= 1){
            return;
        }
        ans += (r-p-1);
        int pivot =  pivot_list[index];
        int id = hashMap.get(pivot);
        solve(p, id);
        solve(id+1, r);
    }
}
